daemon/images/image_import.go

```diff
@@ -3,126 +3,121 @@ package images // import "github.com/docker/docker/daemon/images"
-					return errdefs.InvalidParameter(err)
-				}
-			}
+		if _, isCanonical := newRef.(reference.Canonical); isCanonical {
+			return errdefs.InvalidParameter(errors.New("cannot import digest reference"))
 		}
 
-		config, err := dockerfile.BuildFromConfig(&container.Config{}, changes, os)
-		if err != nil {
-			return err
-		}
-		if src == "-" {
-			rc = inConfig
-		} else {
-			inConfig.Close()
-			if len(strings.Split(src, "://")) == 1 {
-				src = "http://" + src
-			}
-			u, err := url.Parse(src)
+		if tag != "" {
+			_, err = reference.WithTag(newRef, tag)
 			if err != nil {
 				return errdefs.InvalidParameter(err)
 			}
-
-			resp, err = remotecontext.GetWithStatusError(u.String())
-			if err != nil {
-				return err
-			}
-			outStream.Write(streamformatter.FormatStatus("", "Downloading from %s", u))
-			progressOutput := streamformatter.NewJSONProgressOutput(outStream, true)
-			rc = progress.NewProgressReader(resp.Body, progressOutput, resp.ContentLength, "", "Importing")
 		}
+	}
 
-		defer rc.Close()
-		if len(msg) == 0 {
-			msg = "Imported from " + src
+	if src == "-" {
+		rc = inConfig
+	} else {
+		inConfig.Close()
+		if len(strings.Split(src, "://")) == 1 {
+			src = "http://" + src
 		}
-
-		inflatedLayerData, err := archive.DecompressStream(rc)
+		u, err := url.Parse(src)
 		if err != nil {
-			return err
+			return errdefs.InvalidParameter(err)
 		}
-		l, err := i.layerStores[os].Register(inflatedLayerData, "")
-		if err != nil {
-			return err
-		}
-		defer layer.ReleaseAndLog(i.layerStores[os], l)
 
-		created := time.Now().UTC()
-		imgConfig, err := json.Marshal(&image.Image{
-			V1Image: image.V1Image{
-				DockerVersion: dockerversion.Version,
-				Config:        config,
-				Architecture:  runtime.GOARCH,
-				OS:            os,
-				Created:       created,
-				Comment:       msg,
-			},
-			RootFS: &image.RootFS{
-				Type:    "layers",
-				DiffIDs: []layer.DiffID{l.DiffID()},
-			},
-			History: []image.History{{
-				Created: created,
-				Comment: msg,
-			}},
-		})
+		resp, err = remotecontext.GetWithStatusError(u.String())
 		if err != nil {
 			return err
 		}
+		outStream.Write(streamformatter.FormatStatus("", "Downloading from %s", u))
+		progressOutput := streamformatter.NewJSONProgressOutput(outStream, true)
+		rc = progress.NewProgressReader(resp.Body, progressOutput, resp.ContentLength, "", "Importing")
+	}
 
-		// TODO(containerd): Use content store + image store
-		id, err := i.imageStore.Create(imgConfig)
-		if err != nil {
-			return err
-		}
+	defer rc.Close()
+	if len(msg) == 0 {
+		msg = "Imported from " + src
+	}
 
-		// FIXME: connect with commit code and call refstore directly
-		if newRef != nil {
-			if err := i.TagImageWithReference(dgst, newRef); err != nil {
-				return err
-			}
+	inflatedLayerData, err := compression.DecompressStream(rc)
+	if err != nil {
+		return err
+	}
+	defer inflatedLayerData.Close()
+
+	desc, err := archive.ImportIndex(ctx, i.client.ContentStore(), rc)
```

Comment:
For this function is just importing a single layer. It should still register to the layer store for now. The creation of the image artifact will be different, but I think the flow will be

1. register layer with layer store (we can add snapshotter support later), tee output into content store (if not compressed, compress)
2. start containerd lease and create context
3. create image config and put in content store, add label to point to registered layer
4. create manifest from image config and created layer (add gc labels pointing to those artifacts)
5. create image tag
6. release lease

Reply:
Maybe I'm misunderstanding what import index does, but that only partially made sense to me. What is the difference between the index, a layer, and the content? I haven't seen anything about a "containerd lease" either, what is that? I think the differences between docker and containerd are still confusing me.

Reply:
When you import an "index", you are importing a tar in this format https://github.com/opencontainers/image-spec/blob/master/image-layout.md

When you import a "layer", you are importing a tar in this format https://github.com/opencontainers/image-spec/blob/master/layer.md

This is the main difference, but handling those is much different. In the layer case you are basically just given a container filesystem with no other metadata in the tar stream. The other metadata has to be provided separately (in this case in a ton of arguments). Feel free to change the interface to `ImportImage` to make more sense too, it should be called in a limited number of places.

Reply:
> I haven't seen anything about a "containerd lease" either, what is that? I think the differences between docker and containerd are still confusing me.

The containerd interfaces are much lower level than Dockers. In containerd there are two types of "root" objects, images and containers. These root images will link to a set of resources underneath them, such as content blobs or snapshots. Resource management in containerd is strict, meaning any object which is not reference by a root is eligible for garbage collection. Leases can act as a temporary root to reference objects to prevent the garbage collector from removing them before you create your final root object. For example, when creating an image, you will get the content first (whether from a registry or in this case directly from an `io.Reader`), then you will create an image which points to that content. The lease is created before any content is added, then it can ride the context to put any objects created with that context are leased. When the operation is finished, whether successful or not, that lease is removed and any leftover resources are cleaned up. The client then only has to worry about managing a single lease rather than tracking all the resources it may have used or left unused. NOTE: the leases can also have an expiry to handle cases where the client crashes and never comes back or way want to re-attempt the operation before expiration

Reply:
Also see https://github.com/containerd/containerd/blob/master/docs/garbage-collection.md#what-is-a-lease

Reply:
I hadn't found the docs here; this is great, thanks!